overlord/snapstate/handlers.go

```diff
@@ -1934,6 +1934,74 @@ func (m *SnapManager) doUnlinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	return err
 }
 
+func (m *SnapManager) undoUnlinkSnap(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+
+	// XXX: is it safe (can panic)?
+	isInstalled := snapst.IsInstalled()
+	if !isInstalled {
+		return fmt.Errorf("internal error: snap %q not installed", snapsup.InstanceName())
```

Comment:
we can't reach this because to get here we would need succeful discard-snap for the last revision right? but then we would be at the end of remove?

the error should probably say "...: snap %q not anymore installed"

Reply:
Yes, that's right, it's an extra sanity check and therefore I made it "internal error". I'll tweak the message.